### PR TITLE
Tease out reusable elements of Instance, make Networks objecty.

### PR DESCRIPTION
**Pull request by mikalstill**

### CODE REVIEW

File: shakenfist/baseobject.py [hunk 1]
```diff
@@ -0,0 +1,105 @@
...
+    def _db_get(cls, object_uuid):
+        # NOTE(mikal): we don't do upgrades inflight. They are assumed to have been
+        # done as part of the upgrade process.
+        o = etcd.get(cls.object_type, None, object_uuid)
+        if not o:
+            LOG.withField(cls.object_type, object_uuid).error(
+                'Object missing')
+            return None
+
+        if o.get('version', 0) != 2:
```
Comment (mandoonandy) [inline shakenfist/baseobject.py:54]: The 2 should probably be a constant of the class. That would also support a check at the end of `upgrade` to ensure all objects are at the correct version.

File: shakenfist/baseobject.py [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+
+    def _db_set_attribute(self, attribute, value):
+        etcd.put('attribute/%s' % type(self).object_type, self.__uuid, attribute, value)
+
+    # Properties common to all objects which are routed to attributes
+    @property
+    def state(self):
+        return self._db_get_attribute('state')
+
+    def update_state(self, state, error_message=None):
```
Comment (mandoonandy) [inline shakenfist/baseobject.py:73]: It would be good if `update_state` checked that the `state` is valid.

File: shakenfist/baseobject.py (hunk 1 above)
Comment (mikalstill) [reply shakenfist/baseobject.py:54]: Done in 1ec7fa414699eed37c333cf97c6a66e53bbab365.

File: shakenfist/baseobject.py (hunk 2 above)
Comment (mikalstill) [reply shakenfist/baseobject.py:73]: As in the state exists in a whitelist, or that the movement from the original state to this state is valid?

Comment (mandoonandy) [reply shakenfist/baseobject.py:74]: Both. :-)

**Comment by mikalstill** [comment]
CI now passes except for 406 errors, which is nice. The 406s might be masking other failures though. I have put up https://github.com/shakenfist/client-python/pull/22 on python-client to add retries for the 406s, which I need to land before I can progress further here.

File: shakenfist/baseobject.py (hunk 2 above)
Comment (mikalstill) [reply shakenfist/baseobject.py:74]: So I agree, but I think it should be done in a separate change, this one is getting too big.

**Comment by mikalstill** [comment]
Retrying the CI run now that shakenfist/client-python#22 has merged.

**Comment by mikalstill** [comment]
So close... CI's only failure is now a single traceback in the logs. I'm tempted to ask for this to be merged now and then continue on this adventure.

**Comment by mandoonandy** [comment]
Do you want to chase the CI fail? Use the size of the PR as incentive to only chase the CI fail? ;-)

**Comment by mikalstill** [comment]
The final traceback is an attempt by the network daemon to update DHCP for a network which just got deleted.

**Comment by mikalstill** [comment]
Theorem: the rapid reuse of vxids (and therefore interface names) is causing races in the kernel when creating networks. For example in this log of the most recent CI failure:

```
Dec 23 21:13:43 sfci-kg8edv INFO sf-net[12819] Executing ip link add phy-4-o type veth peer name phy-4-i with locks None
Dec 23 21:13:43 sfci-kg8edv INFO sf-net[12819] Executing ip link set phy-4-i netns 11e0ee0f-e9c8-4ebf-9ecb-a2ddd76e16ab with locks None
Dec 23 21:13:43 sfci-kg8edv INFO sf-net[12819] Executing ip netns exec 11e0ee0f-e9c8-4ebf-9ecb-a2ddd76e16ab ip addr show phy-4-i with locks None
Dec 23 21:13:43 sfci-kg8edv INFO sf-net[12819] Executing ip netns exec 11e0ee0f-e9c8-4ebf-9ecb-a2ddd76e16ab ip addr add 192.168.20.134/255.255.255.0 dev phy-4-i with locks None
Dec 23 21:13:43 sfci-kg8edv INFO sf-net[12819] Executing ip netns exec 11e0ee0f-e9c8-4ebf-9ecb-a2ddd76e16ab ip link set phy-4-i up with locks None
Dec 23 21:13:43 sfci-kg8edv INFO sf-net[12819] Executing ip netns exec 11e0ee0f-e9c8-4ebf-9ecb-a2ddd76e16ab iptables -A FORWARD -o phy-4-i -i veth-4-i -j ACCEPT with locks None
Dec 23 21:13:43 sfci-kg8edv INFO sf-net[12819] Executing ip netns exec 11e0ee0f-e9c8-4ebf-9ecb-a2ddd76e16ab iptables -A FORWARD -i phy-4-i -o veth-4-i -j ACCEPT with locks None
Dec 23 21:13:43 sfci-kg8edv INFO sf-net[12819] Executing ip netns exec 11e0ee0f-e9c8-4ebf-9ecb-a2ddd76e16ab iptables -t nat -A POSTROUTING -s 192.168.242.0/255.255.255.0 -o phy-4-i -j MASQUERADE with locks None
Dec 23 21:14:18 sfci-kg8edv DEBUG gunicorn: worker [sf-api][12933] net.delete(); broadcast=192.168.242.255; in_netns=ip netns exec 11e0ee0f-e9c8-4ebf-9ecb-a2ddd76e16ab; ipblock=192.168.242.0; netmask=255.255.255.0; netns=11e0ee0f-e9c8-4ebf-9ecb-a2ddd76e16ab; physical_bridge=phy-br-ens4; physical_interface=ens4; physical_veth_inner=phy-4-i; physical_veth_outer=phy-4-o; router=192.168.242.1; vx_bridge=br-vxlan-4; vx_id=4; vx_interface=vxlan-4; vx_veth_inner=veth-4-i; vx_veth_outer=veth-4-o
Dec 23 21:14:24 sfci-kg8edv INFO sf-net[12819] Executing ip link add phy-4-o type veth peer name phy-4-i with locks None
Dec 23 21:14:25 sfci-kg8edv INFO sf-net[12819] Executing ip link set phy-4-i netns c71230c6-8b9e-4c1e-b822-fd3d108765de with locks None
Dec 23 21:14:25 sfci-kg8edv INFO sf-net[12819] Executing ip netns exec c71230c6-8b9e-4c1e-b822-fd3d108765de ip addr show phy-4-i with locks None
Dec 23 21:14:25 sfci-kg8edv INFO sf-net[12819] Executing ip netns exec c71230c6-8b9e-4c1e-b822-fd3d108765de ip addr add 192.168.20.74/255.255.255.0 dev phy-4-i with locks None
Dec 23 21:14:25 sfci-kg8edv INFO sf-net[12819] Executing ip netns exec c71230c6-8b9e-4c1e-b822-fd3d108765de ip link set phy-4-i up with locks None
Dec 23 21:14:25 sfci-kg8edv kernel: [  642.130894] IPv6: ADDRCONF(NETDEV_CHANGE): phy-4-i: link becomes ready
Dec 23 21:14:25 sfci-kg8edv INFO sf-net[12819] Executing ip netns exec c71230c6-8b9e-4c1e-b822-fd3d108765de iptables -A FORWARD -o phy-4-i -i veth-4-i -j ACCEPT with locks None
Dec 23 21:14:25 sfci-kg8edv INFO sf-net[12819] Executing ip netns exec c71230c6-8b9e-4c1e-b822-fd3d108765de iptables -A FORWARD -i phy-4-i -o veth-4-i -j ACCEPT with locks None
Dec 23 21:14:25 sfci-kg8edv INFO sf-net[12819] Executing ip netns exec c71230c6-8b9e-4c1e-b822-fd3d108765de iptables -t nat -A POSTROUTING -s 192.168.242.0/255.255.255.0 -o phy-4-i -j MASQUERADE with locks None
Dec 23 21:15:54 sfci-kg8edv DEBUG sf-queues-1608758152.219226-000[22297] net.delete(); broadcast=192.168.242.255; in_netns=ip netns exec c71230c6-8b9e-4c1e-b822-fd3d108765de; ipblock=192.168.242.0; netmask=255.255.255.0; netns=c71230c6-8b9e-4c1e-b822-fd3d108765de; physical_bridge=phy-br-ens4; physical_interface=ens4; physical_veth_inner=phy-4-i; physical_veth_outer=phy-4-o; router=192.168.242.1; vx_bridge=br-vxlan-4; vx_id=4; vx_interface=vxlan-4; vx_veth_inner=veth-4-i; vx_veth_outer=veth-4-o
Dec 23 21:15:56 sfci-kg8edv DEBUG gunicorn: worker [sf-api][12933] net.delete(); broadcast=192.168.242.255; in_netns=ip netns exec c71230c6-8b9e-4c1e-b822-fd3d108765de; ipblock=192.168.242.0; netmask=255.255.255.0; netns=c71230c6-8b9e-4c1e-b822-fd3d108765de; physical_bridge=phy-br-ens4; physical_interface=ens4; physical_veth_inner=phy-4-i; physical_veth_outer=phy-4-o; router=192.168.242.1; vx_bridge=br-vxlan-4; vx_id=4; vx_interface=vxlan-4; vx_veth_inner=veth-4-i; vx_veth_outer=veth-4-o
Dec 23 21:15:58 sfci-kg8edv INFO sf-net[12819] Executing ip link add phy-4-o type veth peer name phy-4-i with locks None
Dec 23 21:15:58 sfci-kg8edv INFO sf-net[12819] Executing ip link set phy-4-i netns 82bfe74b-0b3b-4151-8d09-be2c75086d77 with locks None
Dec 23 21:15:58 sfci-kg8edv INFO sf-net[12819] Executing ip netns exec 82bfe74b-0b3b-4151-8d09-be2c75086d77 ip addr show phy-4-i with locks None
Dec 23 21:15:58 sfci-kg8edv INFO sf-net[12819] Executing ip netns exec 82bfe74b-0b3b-4151-8d09-be2c75086d77 ip addr add 192.168.20.123/255.255.255.0 dev phy-4-i with locks None
Dec 23 21:15:58 sfci-kg8edv INFO sf-net[12819] Executing ip netns exec 82bfe74b-0b3b-4151-8d09-be2c75086d77 ip link set phy-4-i up with locks None
Dec 23 21:15:59 sfci-kg8edv INFO sf-net[12819] Executing ip netns exec 82bfe74b-0b3b-4151-8d09-be2c75086d77 iptables -A FORWARD -o phy-4-i -i veth-4-i -j ACCEPT with locks None
Dec 23 21:15:59 sfci-kg8edv INFO sf-net[12819] Executing ip netns exec 82bfe74b-0b3b-4151-8d09-be2c75086d77 iptables -A FORWARD -i phy-4-i -o veth-4-i -j ACCEPT with locks None
Dec 23 21:15:59 sfci-kg8edv INFO sf-net[12819] Executing ip netns exec 82bfe74b-0b3b-4151-8d09-be2c75086d77 iptables -t nat -A POSTROUTING -s 192.168.242.0/255.255.255.0 -o phy-4-i -j MASQUERADE with locks None
Dec 23 21:16:00 sfci-kg8edv INFO sf-net[12819] Executing ip netns exec 5b3bdd6d-50e6-4e45-b748-cb312cdbf1ec ip addr show phy-4-i with locks None
Dec 23 21:16:00 sfci-kg8edv INFO sf-net[12819] Executing ip netns exec 5b3bdd6d-50e6-4e45-b748-cb312cdbf1ec ip addr add 192.168.20.135/255.255.255.0 dev phy-4-i with locks None
Dec 23 21:16:00 sfci-kg8edv sh[12685]: Command: ip netns exec 5b3bdd6d-50e6-4e45-b748-cb312cdbf1ec ip addr add 192.168.20.135/255.255.255.0 dev phy-4-i
Dec 23 21:16:00 sfci-kg8edv sh[12685]: Stderr: 'Cannot find device "phy-4-i"\n'
```

You can see we use the vxid 4 over and over, and the flow works well the first two times and then doesn't. Perhaps we should randomise vxids like we do IP allocations in IPManager...

**Comment by mandoonandy** [comment]
I have often thought that was a possibility. Rather than have random race conditions, maybe use incremental id's thus guaranteeing no race conditions.

**Review by mandoonandy — APPROVED** [review]
LGTM

**Comment by mikalstill** [comment]
> I have often thought that was a possibility. Rather than have random race conditions, maybe use incremental id's thus guaranteeing no race conditions.

IPs are allocated randomly to reduce the chance of someone guessing one. I think I like the same idea here. If you owned our network traffic somehow, not being able to predict what vxid a given network would have is useful.

**Pull request merged**